Improve setup.py and testing

cdeil opened the pull request:
This PR does the following changes:
- [x] Declare dependencies other than Numpy and Astropy as optional in `setup.py`.
- [x] Add a Python 3 build to `.travis.yml` with all dependencies present except ST.
- [x] Add a Python 2 build to `.travis.yml` with all dependencies present except ST.
- [x] Add a Python 3 build to `.travis.yml` with only Numpy / Astropy present, i.e. minimal dependencies (doesn't pass yet, commented out)
- [x] Add `fermipy/tests/test_skymap.py`
- [x] Add `requires_dependency` decorators to tests, just like we do in Gammapy.
- [x] Misc code cleanup, especially in tests.

This should make `pip install fermipy` work for more users and developers and help a little with testing / interoperability with Gammapy.

cdeil commented:
@woodmd - Can you please review?

woodmd commented on fermipy/plotting.py:
```diff
@@ -1,15 +1,7 @@
-except KeyError:
-    matplotlib.use('Agg')
-
 import matplotlib.pyplot as plt
```
What's the motivation for removing this?  I've found it useful if you want plotting to generate plots on a machine without an X server (e.g. the nodes on our computing cluster at SLAC).  Is there another way to accomplish the same behavior?

cdeil replied:
Usually configuring the matplotlib backend is something that should be left up to the end user, not done automatically by libraries.

For non-interactive plotting, what you do here is the right thing to do:
```python
import matplotlib
matplotlib.use('agg')
```

I have that (and other plotting configuration) at the top of every non-interactive script that makes plots.
This can be explained to users in the docs.

Doing it automatically and sometimes on `import fermipy.plotting` can be confusing or annoying.
So I'm +1 to remove this. But I don't feel strongly, if you like it that way let me know and I'll put it back.

woodmd commented:
Why did you drop the `unicode_literals` import?  I thought this was needed for python3 compatibility?

cdeil commented:
> Why did you drop the unicode_literals imports? I thought this was needed for python3 compatibility?

It's very controversial if `from __future__ import unicode_literals` is a good or bad idea.
http://python-future.org/unicode_literals.html
https://docs.python.org/3/howto/pyporting.html#text-versus-binary-data

It's possible to achieve a single code base that works on Py 2 and 3 with or without it.

Here I removed it because I got weird errors from the SWIG C extensions when I added it, and didn't have time to investigate or fix the issues.

When it comes to strings, `fermipy` is not yet fully Python 3 compatible.
There's several things in e.g. https://github.com/fermiPy/fermipy/blob/master/fermipy/utils.py that need to be changed, e.g. `basestring` doesn't exist in Python 3.

Can we defer this to future PRs?
The best way to do it is to add tests that show what is / isn't working, and then to make those tests work on Python 2 and 3.

cdeil commented:
One more thing: my understanding is that `from __future__ import unicode_literals` is a no-op in Python 3, it changes the behaviour of Python 2 to be more like Python 3.
So not adding this import to Fermipy preserves the status quo: a working codebase in Python 2.

woodmd commented:
Ok I take your point regarding the matplotlib backend.  Maybe we can just create a utility function that can be called if the user wants to automatically fall back to Agg.

Regarding the unicode conversions, getting rid of this would definitely simplify maintaining the ST-facing part of the code.  When unicode strings are enabled all strings passed to SWIG methods need to be wrapped with `str` but I often forget to do this.  Deferring the python3 compatibility issues to a future PR is fine with me.
